### ISSUE TITLE
Feature request: instance variable checking

### ISSUE THREAD

**Issue by ConradIrwin** [za-creature/coffeescope#27]
### Code sample

``` coffeescript
class User
  constructor: (@_name) ->

  getName: ->
     @_capitalize(@_naem)

  _capitalizeWord: (word) ->
     word.toUpperCase()
```

It would be nice to catch that `getName` is broken because neither `_naem` nor `_capitalize` are defined.

This could be done by starting with a scope of "inherited methods", e.g.:

```
# coffeelint.json

  "inherited": {
     "React.Component": {
          "render": true, // render is inherited from 'React.Component'
          "state": false,
          "setState": true
    }
  }
```

And then adding all instance methods defined on the current class to that scope, and also all variables mentioned in the constructor.

It is good practice in javascript (and thus coffeescript) to initialize all instance variables in your constructor to avoid the object changing it's ["hidden class"](https://developers.google.com/v8/design) at runtime.

I'd love to use a tool like this, and would be happy to try and build it. Would you like it as part of coffeescope2, or should I make my own coffeelint plugin?

**Comment by za-creature** [comment]
I've given it some thought, and at the end of the day I have to say that I'm against this landing in coffeescope2. That being said, I'm always willing to reconsider if you (or someone else) makes a good argument for it.

On one hand, it'd be awesome to have this however, what coffeescope2 has been doing thus far has been pretty deterministic. Most options are toggles and / or exceptions (i.e. don't complain about accessing the `Math` global object), but the errors coffeescope2 does report are for the most part real problems (unless I totally fuck up the implementation in which case they're false positives)

What you are suggesting is in the realm of static typing and can't reliably be implemented in a dynamic language like coffeescript (or python for that matter). I could _kind-of_ see this working with [.d.ts files](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/jquery/jquery.d.ts) but at that point, you're probably better off just writing typescript instead.

Here are a few issues:
- Keeping track of scope in the presence of `->` and `=>` and `Function::apply` and `Function::call`
- Figuring out if a property is ever assigned to in the body of a class or any ancestors (which could exist in different files or potentially even different compile-to-js languages)
- Reliably determining `no-use-before-define` in the same context and in the presence of `super`
  
  ``` coffeescript
  console.log(@name)
  @name = "foo"
  ```

TL;DR; You should make your own, but I'm not sure if a coffeelint plugin is what you should go for; you're looking for something that's a bit more project-centric and coffeelint (and coffeescope) for the most part deal with files.

Going to leave it open for now in the event anyone has a case to make about landing this in coffeescope, but I personally don't see it happening; it's like trying to attach a rocket engine to a bicycle.

**Comment by za-creature** [comment]
That being said, [I am up for coffeelint2](https://github.com/za-creature/coffeescope/issues/21#issuecomment-226495388). Wanna make this an org?

**Comment by za-creature** [comment]
Time's up, closing
